File: src/main/java/de/bnder/taskmanager/commands/Stats.java
package de.bnder.taskmanager.commands;

import com.google.cloud.firestore.DocumentSnapshot;
import de.bnder.taskmanager.main.Command;
import de.bnder.taskmanager.main.Main;
import de.bnder.taskmanager.utils.Localizations;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;

import java.awt.*;
import java.util.Date;
import java.util.concurrent.ExecutionException;
import java.util.Locale;

public class Stats implements Command {

    @Override
    public void action(String[] args, String messageContentRaw, Member commandExecutor, TextChannel textChannel, Guild guild, java.util.List<Member> mentionedMembers, java.util.List<Role> mentionedRoles, java.util.List<TextChannel> mentionedChannels, SlashCommandEvent slashCommandEvent) {
        guild.getJDA().getRestPing().queue(ping -> {
            final Locale langCode = Localizations.getGuildLanguage(guild);
            final EmbedBuilder builder = new EmbedBuilder();
            try {
                final Date apiRequestDateStart = new Date();
                DocumentSnapshot getStats = Main.firestore.collection("stats").document("alltime").get().get();
                final Date apiRequestDateEnd = new Date();

                long serversShard0 = 0;
                if (getStats.exists() && getStats.getData().containsKey("servers_shard_0")) {
                    serversShard0 = (long) getStats.get("servers_shard_0");
                }
                long serversShard1 = 0;
                if (getStats.exists() && getStats.getData().containsKey("servers_shard_1")) {
                    serversShard1 = (long) getStats.get("servers_shard_1");
                }
                long tasksDone = 0;
                if (getStats.exists() && getStats.getData().containsKey("tasks_done")) {
                    tasksDone = (long) getStats.get("tasks_done");
                }
                long tasksCreated = 0;
                if (getStats.exists() && getStats.getData().containsKey("tasks_created")) {
                    tasksCreated = (long) getStats.get("tasks_created");
                }
                long messagesSent = 0;
                if (getStats.exists() && getStats.getData().containsKey("messages_sent")) {
                    messagesSent = (long) getStats.get("messages_sent");
                }

                builder.addField(Localizations.getString("stats_field_server", langCode), String.valueOf(serversShard0 + serversShard1), true);
                builder.addField(Localizations.getString("stats_field_tasks_done", langCode), String.valueOf(tasksDone), true);
                builder.addField(Localizations.getString("stats_field_tasks_created", langCode), String.valueOf(tasksCreated), true);
                builder.addField("Shard", String.valueOf((guild.getJDA().getShardInfo().getShardId())), true);
                builder.addField("Shards", String.valueOf(guild.getJDA().getShardInfo().getShardTotal()), true);
                builder.addField(Localizations.getString("stats_field_server", langCode) + " Shard 0", String.valueOf(serversShard0), true);
                builder.addField(Localizations.getString("stats_field_server", langCode) + " Shard 1", String.valueOf(serversShard1), true);
                builder.addField(Localizations.getString("stats_field_messages_sent", langCode), String.valueOf(messagesSent), true);
                builder.addField(Localizations.getString("stats_field_ping_discord", langCode), ping + "ms", true);
                builder.addField(Localizations.getString("stats_field_ping_bnder", langCode), (apiRequestDateEnd.getTime() - apiRequestDateStart.getTime()) + "ms", true);
            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
            builder.setColor(Color.green);
            builder.setTitle(Localizations.getString("stats_message_title", langCode));
            handleEmbedsOnSlashCommand(textChannel, slashCommandEvent, builder);
        });
    }

    public static void handleEmbedsOnSlashCommand(TextChannel textChannel, SlashCommandEvent slashCommandEvent, EmbedBuilder builder) {
        if (slashCommandEvent == null) {
            textChannel.sendMessageEmbeds(builder.build()).queue();
        } else {
            StringBuilder reply = new StringBuilder();
            for (MessageEmbed.Field field : builder.getFields()) {
                reply.append(field.getName()).append(": ").append(field.getValue()).append("\n");
            }
            slashCommandEvent.reply(reply.toString()).queue();
        }
    }
}

File: src/main/java/de/bnder/taskmanager/commands/Stats.kt
package de.bnder.taskmanager.commands

import com.eclipsesource.json.Json
import de.bnder.taskmanager.main.Command
import de.bnder.taskmanager.main.Main
import de.bnder.taskmanager.utils.Connection
import de.bnder.taskmanager.utils.Localizations
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.jsoup.Jsoup
import java.awt.Color
import java.io.IOException

/*
 * Copyright (C) 2019 Jan Brinkmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class Stats : Command {
    override fun action(args: Array<String>, event: GuildMessageReceivedEvent) {
        val server = event.jda.guilds.size
        val ping = event.jda.gatewayPing
        val langCode = Localizations.getGuildLanguage(event.guild)
        val builder = EmbedBuilder()
        builder.addField(Localizations.getString("stats_field_server", langCode), server.toString(), true)
        builder.addField(Localizations.getString("stats_field_ping", langCode), ping.toString(), true)
        try {
            val res = Jsoup.connect(Main.requestURL + "/stats").method(org.jsoup.Connection.Method.GET).header("authorization", "TMB " + Main.authorizationToken).header("user_id", event.member!!.id).timeout(Connection.timeout).userAgent(Main.userAgent).ignoreContentType(true).ignoreHttpErrors(true).execute()
            val jsonObject = Json.parse(res.parse().body().text()).asObject()
            builder.addField(Localizations.getString("stats_field_tasks_done", langCode), jsonObject.getInt("tasks_done", -1).toString(), true)
            builder.addField(Localizations.getString("stats_field_tasks_created", langCode), jsonObject.getInt("tasks_created", -1).toString(), true)
            builder.addField("Shard", (event.jda.shardInfo.shardId).toString(), true)
            builder.addField("Shards", event.jda.shardInfo.shardTotal.toString(), true)
            builder.addField(Localizations.getString("stats_field_requests_received", langCode), jsonObject.getInt("received_requests", -1).toString(), true)
            builder.addField(Localizations.getString("stats_field_messages_sent", langCode), jsonObject.getInt("messages_sent", -1).toString(), true)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        builder.setColor(Color.green)
        builder.setTitle(Localizations.getString("stats_message_title", langCode))
        event.channel.sendMessage(builder.build()).queue()
    }
}